README.md
<h1 align="center">CosmiQ Works Evaluation Tools</h1>
<p align="center">
<a href="http://www.cosmiqworks.org"><img src="http://www.cosmiqworks.org/wp-content/uploads/2016/02/cropped-CosmiQ-Works-Logo_R_RGB.png" width="350" alt="CosmiQ Works"></a>
<br>
<br>
<img align="center" src="https://img.shields.io/pypi/pyversions/cw-eval.svg">
<img align="center" src="https://img.shields.io/conda/vn/conda-forge/cw-eval.svg">
<img align="center" src="https://readthedocs.org/projects/pip/badge/" alt="docs">
<img align="center" src="https://img.shields.io/github/license/cosmiq/cw-eval.svg" alt="license">
<img align="center" src="https://img.shields.io/docker/build/cosmiqworks/cw-eval.svg" alt="docker">
</p>

- [Installation Instructions](#installation-instructions)
- [API Documentation](https://cw-eval.readthedocs.io/)
- [Dependencies](#dependencies)
- [License](#license)
- [Download Instructions](#spacenet-dataset-download-instructions)
---
Modeled off of capabilities from [rio-tiler](https://github.com/mapbox/rio-tiler) and designed for use with SpaceNet datasets, this library provides code for tiling geospatial imagery datasets into manageable pieces.


## Installation Instructions
Several packages require binaries to be installed before pip installing the other packages.  Conda is a simple way to install everything and their dependencies:

#### Conda
_note: This functionality is not ready as of 12.11.18. Check back soon._
```
conda install -c conda-forge cw-tiler
```

#### pip

You may use `pip` to install this package; however, note that one of the dependencies, [rtree](https://github.com/Toblerity/rtree), can require pre-installation of [libspatialindex](https://libspatialindex.github.io/) binaries. This can all be done by installing rtree using conda:
```
conda install -c conda-forge rtree
```
or by [following the instructions for libspatialindex installation](https://libspatialindex.github.io/).

Once you have dependencies set up, install as usual using `pip`:
```
pip install cw-tiler
```
For bleeding-edge versions (use at your own risk), `pip install` from the dev branch of this repository:
```
pip install --upgrade git+https://github.com/CosmiQ/cw-tiler.git@dev
```

#### Docker

You may also use our Docker container:
```
docker pull cosmiqworks/cw-tiler
```

## API Documentation
See the [readthedocs](https://cw-tiler.readthedocs.io/) page.

## Dependencies
All dependencies can be found in the docker file [Dockerfile](./Dockerfile) or
[environment.yml](./environment.yml)

## License
See [LICENSE](./LICENSE.txt).

## SpaceNet Dataset Download Instructions
Further download instructions for the [SpaceNet Dataset](https://github.com/SpaceNetChallenge/utilities/tree/master/content/download_instructions) can be found [here](https://github.com/SpaceNetChallenge/utilities/tree/master/content/download_instructions).


CHANGES.txt
0.0.1 (2018-6-12)
------------------
- Initial release. Requires rio-tiler


tests/test_main_cv.py
"""tests rio_tiler.base"""

import os
import pytest
import rasterio
import random
from cw_tiler import main
from cw_tiler import utils
import numpy as np


utmX, utmY = 658029, 4006947
ll_x = utmX
ur_x = utmX + 500
ll_y = utmY
ur_y = utmY + 500
stride_size_meters = 300
cell_size_meters = 400
tile_size_pixels = 1600
spacenetPath = "s3://spacenet-dataset/AOI_2_Vegas/srcData/rasterData/AOI_2_Vegas_MUL-PanSharpen_Cloud.tif"

PREFIX = os.path.join(os.path.dirname(__file__), 'fixtures')
ADDRESS = '{}/my-bucket/hro_sources/colorado/201404_13SED190110_201404_0x1500m_CL_1.tif'.format(PREFIX)
ADDRESS_ALPHA = '{}/my-bucket/hro_sources/colorado/201404_13SED190110_201404_0x1500m_CL_1_alpha.tif'.format(PREFIX)
ADDRESS_NODATA = '{}/my-bucket/hro_sources/colorado/201404_13SED190110_201404_0x1500m_CL_1_nodata.tif'.format(PREFIX)


def test_calculate_utm_epsg():

    wgs_bounds = utils.get_wgs84_bounds(ADDRESS)
    print(wgs_bounds)
    utm_EPSG = utils.calculate_UTM_EPSG(wgs_bounds)

    assert utm_EPSG=="EPSG:32613"


def test_grid_creation():

    address = spacenetPath

    with rasterio.open(address) as src:

        wgs_bounds = utils.get_wgs84_bounds(src)
        utm_EPSG = utils.calculate_UTM_EPSG(wgs_bounds)
        utm_bounds = utils.get_utm_bounds(src, utm_EPSG)
        print(utm_bounds)
        cells_list = main.calculate_analysis_grid(utm_bounds, stride_size_meters=stride_size_meters, cell_size_meters=cell_size_meters)
        print(len(cells_list))

    assert len(cells_list[0]) == 4
    assert ((cells_list[0][2]-cells_list[0][0]), (cells_list[0][3]-cells_list[0][1])) == (cell_size_meters, cell_size_meters)
    assert (cells_list[1][1] - cells_list[0][1]) == stride_size_meters
    assert len(cells_list) == 2592




def test_return_tile():

    address = spacenetPath

    with rasterio.open(address) as src:

        wgs_bounds = utils.get_wgs84_bounds(src)
        utm_EPSG = utils.calculate_UTM_EPSG(wgs_bounds)
        utm_bounds = utils.get_utm_bounds(src, utm_EPSG)

        cells_list = main.calculate_analysis_grid(utm_bounds, stride_size_meters=stride_size_meters, cell_size_meters=cell_size_meters)

        random_cell = random.choice(cells_list)
        ll_x, ll_y, ur_x, ur_y = random_cell
        tile, mask = main.tile_utm(src, ll_x, ll_y, ur_x, ur_y, indexes=None, tilesize=tile_size_pixels, nodata=None, alpha=None,
                 dst_crs=utm_EPSG)

        print(np.shape(tile))
        assert np.shape(tile) == (8, tile_size_pixels, tile_size_pixels)








cw_tiler/utils.py
"""cv_tiler.utils: utility functions."""


import numpy as np
import rasterio
from rasterio.vrt import WarpedVRT
from rasterio.enums import Resampling
from rasterio.io import DatasetReader
from rasterio.warp import transform_bounds
from rio_tiler.errors import RioTilerError

from shapely.geometry import box
import statistics

from PIL import Image

# Python 2/3
try:
    from urllib.request import urlopen
except ImportError:
    from urllib2 import urlopen


def utm_getZone(longitude):
    """Calculate UTM Zone from Longitude

    Attributes:
    ___________
    longitude: float of longitude (Degrees.decimal degrees)

    Returns:
    _______
    out: int
        returns UTM Zone number
    """
    return (int(1+(longitude+180.0)/6.0))

def utm_isNorthern(latitude):
    """Calculate UTM North/South from Latitude

    Attributes:
    ___________
    latitude: float of latitude (Deg.decimal degrees)

    Returns:
    _______
    out: int
        returns UTM Zone number
    """

    if (latitude < 0.0):
        return "S"
    else:
        return "N"


def calculate_UTM_EPSG(coords, epsgStart={"N":"EPSG:326", "S":"EPSG:327"}):
    """Calculate UTM North/South from Latitude

    Attributes:
    ___________
    coords: list, [longitude, latitude] or [min_longitude, min_latitude, max_lognitude, max_latitude
    epsgStart: dict, dictionary for precursor for UTM Zone EPSG (Default is WGS84)

    Returns:
    _______
    out: str
        returns UTM Zone EPSG
    """
    if len(coords) == 2:
        longitude, latitude = coords
    elif len(coords) == 4:
        longitude = statistics.mean([coords[0], coords[2]])
        latitude = statistics.mean([coords[1], coords[3]])

    utm_zone = utm_getZone(longitude)

    utm_direction = utm_isNorthern(latitude)


    return "{}{}".format(epsgStart[utm_direction], str(utm_zone).zfill(2))

def tile_read_utm(source, bounds, tilesize, indexes=[1], nodata=None, alpha=None, dst_crs='EPSG:3857'):
    """Read data and mask

    Attributes
    ----------
    source : str or rasterio.io.DatasetReader
        input file path or rasterio.io.DatasetReader object
    bounds : tuple, (w, s, e, n) bounds in dst_csrs
    tilesize : Output image size
    indexes : list of ints or a single int, optional, (default: 1)
        If `indexes` is a list, the result is a 3D array, but is
        a 2D array if it is a band index number.
    nodata: int or float, optional (defaults: None)
    alpha: int, optional (defaults: None)
        Force alphaband if not present in the dataset metadata

    dst_crs: str, optional (defaults: "EPSG:3857" (Web Mercator)
        Determine output path


    Returns
    -------
    out : array, int
        returns pixel value.
    """
    w, s, e, n = bounds

    if alpha is not None and nodata is not None:
        raise RioTilerError('cannot pass alpha and nodata option')

    if isinstance(indexes, int):
        indexes = [indexes]

    out_shape = (len(indexes), tilesize, tilesize)

    vrt_params = dict(
        dst_crs=dst_crs,
        resampling=Resampling.bilinear,
        src_nodata=nodata,
        dst_nodata=nodata)

    if isinstance(source, DatasetReader):
        with WarpedVRT(source, **vrt_params) as vrt:
            window = vrt.window(w, s, e, n, precision=21)
            data = vrt.read(window=window,
                            boundless=True,
                            resampling=Resampling.bilinear,
                            out_shape=out_shape,
                            indexes=indexes)

            if nodata is not None:
                mask = np.all(data != nodata, axis=0).astype(np.uint8) * 255
            elif alpha is not None:
                mask = vrt.read(alpha, window=window,
                                out_shape=(tilesize, tilesize),
                                boundless=True,
                                resampling=Resampling.bilinear)
            else:
                mask = vrt.read_masks(1, window=window,
                                      out_shape=(tilesize, tilesize),
                                      boundless=True,
                                      resampling=Resampling.bilinear)
    else:
        with rasterio.open(source) as src:
            with WarpedVRT(src, **vrt_params) as vrt:
                window = vrt.window(w, s, e, n, precision=21)
                data = vrt.read(window=window,
                                boundless=True,
                                resampling=Resampling.bilinear,
                                out_shape=out_shape,
                                indexes=indexes)

                if nodata is not None:
                    mask = np.all(data != nodata, axis=0).astype(np.uint8) * 255
                elif alpha is not None:
                    mask = vrt.read(alpha, window=window,
                                    out_shape=(tilesize, tilesize),
                                    boundless=True,
                                    resampling=Resampling.bilinear)
                else:
                    mask = vrt.read_masks(1, window=window,
                                          out_shape=(tilesize, tilesize),
                                          boundless=True,
                                          resampling=Resampling.bilinear)

    return data, mask

def tile_exists_utm(boundsSrc, boundsTile):
    """"Check if suggested tile is within bounds

    'bounds = (w, s, e, n)'
    'box( minx, miny, maxx, maxy)'

    :param bounds:

    :return:
    """


    boundsSrcBox = box(*boundsSrc)
    boundsTileBox = box(*boundsTile)

    return boundsSrcBox.intersects(boundsTileBox)




def get_wgs84_bounds(source):
    if isinstance(source, DatasetReader):
        src = source

        wgs_bounds = transform_bounds(*[src.crs, 'epsg:4326'] + list(src.bounds), densify_pts=21)

    else:
        with rasterio.open(source) as src:

            wgs_bounds = transform_bounds(*[src.crs, 'epsg:4326'] + list(src.bounds), densify_pts=21)

    return wgs_bounds


def get_utm_bounds(source, utm_EPSG):
    if isinstance(source, DatasetReader):
        src = source

        utm_bounds = transform_bounds(*[src.crs, utm_EPSG] + list(src.bounds), densify_pts=21)

    else:
        with rasterio.open(source) as src:

            utm_bounds = transform_bounds(*[src.crs, utm_EPSG] + list(src.bounds), densify_pts=21)

    return utm_bounds





cw_tiler/main.py
import rasterio
from rasterio.warp import transform_bounds
from rasterio.io import DatasetReader
import statistics
import math
import random
from rio_tiler.errors import TileOutsideBounds
from cw_tiler import utils
import numpy as np



def tile_utm_source(src, ll_x, ll_y, ur_x, ur_y, indexes=None, tilesize=256, nodata=None, alpha=None, dst_crs='epsg:4326'):
    """
    Create UTM tile from any images.

    Attributes
    ----------
    source : rasterio.Dataset
    tile_x : int
        Mercator tile X index.
    tile_y : int
        Mercator tile Y index.
    tile_z : int
        Mercator tile ZOOM level.
    indexes : tuple, int, optional (default: (1, 2, 3))
        Bands indexes for the RGB combination.
    tilesize : int, optional (default: 256)
        Output image size.
    nodata: int or float, optional
        Overwrite nodata value for mask creation.
    alpha: int, optional
        Overwrite alpha band index for mask creation.


    Returns
    -------
    data : numpy ndarray
    mask: numpy array

    """

    wgs_bounds = transform_bounds(
        *[src.crs, dst_crs] + list(src.bounds), densify_pts=21)

    indexes = indexes if indexes is not None else src.indexes
    tile_bounds = (ll_x, ll_y, ur_x, ur_y)
    if not utils.tile_exists_utm(wgs_bounds, tile_bounds):
        raise TileOutsideBounds(
            'Tile {}/{}/{}/{} is outside image bounds'.format(ll_x, ll_y, ur_x, ur_y))

    return utils.tile_read_utm(src,
                               tile_bounds,
                               tilesize,
                               indexes=indexes,
                               nodata=nodata,
                               alpha=alpha,
                               dst_crs=dst_crs)



def tile_utm(source, ll_x, ll_y, ur_x, ur_y, indexes=None, tilesize=256, nodata=None, alpha=None, dst_crs='epsg:4326'):
    """
    Create UTM tile from any images.

    Attributes
    ----------
    address : str
        file url or rasterio.Dataset.
    tile_x : int
        Mercator tile X index.
    tile_y : int
        Mercator tile Y index.
    tile_z : int
        Mercator tile ZOOM level.
    indexes : tuple, int, optional (default: (1, 2, 3))
        Bands indexes for the RGB combination.
    tilesize : int, optional (default: 256)
        Output image size.
    nodata: int or float, optional
        Overwrite nodata value for mask creation.
    alpha: int, optional
        Overwrite alpha band index for mask creation.


    Returns
    -------
    data : numpy ndarray
    mask: numpy array

    """

    if isinstance(source, DatasetReader):

        return tile_utm_source(source,
                               ll_x,
                               ll_y,
                               ur_x,
                               ur_y,
                               indexes=indexes,
                               tilesize=tilesize,
                               nodata=nodata,
                               alpha=alpha,
                               dst_crs=dst_crs)



    else:
        with rasterio.open(source) as src:

            return tile_utm_source(src,
                                   ll_x,
                                   ll_y,
                                   ur_x,
                                   ur_y,
                                   indexes=indexes,
                                   tilesize=tilesize,
                                   nodata=nodata,
                                   alpha=alpha,
                                   dst_crs=dst_crs)




def get_chip(source, ll_x, ll_y, xres, yres,
                 utm_EPSG=[],
                 indexes=None,
             tile_size_pixels=512,
                 nodata=None,
                 alpha=None):
    """ Get Chip From Image Address"

    :param address:
    :param ll_x:
    :param ll_y:
    :param xres:
    :param yres:
    :param downSampleRate:
    :return:

    """

    if isinstance(source, DatasetReader):

        if not utm_EPSG:
            wgs_bounds = utils.get_wgs84_bounds(source)
            utm_EPSG = utils.calculate_UTM_EPSG(wgs_bounds)

        return  tile_utm(source, ll_x, ll_y, ur_x, ur_y,
                               indexes=indexes,
                               tilesize=tile_size_pixels,
                               nodata=nodata,
                               alpha=alpha,
                 dst_crs=utm_EPSG)



    else:
        with rasterio.open(source) as src:

            wgs_bounds = utils.get_wgs84_bounds(src)
            utm_EPSG = utils.calculate_UTM_EPSG(wgs_bounds)

            return tile_utm(source, ll_x, ll_y, ur_x, ur_y,
                     indexes=indexes,
                     tilesize=tile_size_pixels,
                     nodata=nodata,
                     alpha=alpha,
                     dst_crs=utm_EPSG)





def calculate_anchor_points(utm_bounds, stride_size_meters=400):
    min_x = math.floor(utm_bounds[0])
    min_y = math.floor(utm_bounds[1])
    max_x = math.ceil(utm_bounds[2])
    max_y = math.ceil(utm_bounds[3])

    anchor_point_list = []
    for x in np.arange(min_x, max_x, stride_size_meters):
        for y in np.arange(min_y, max_y, stride_size_meters):
            anchor_point_list.append([x, y])


    return anchor_point_list


def calculate_cells(anchor_point_list, cell_size_meters):
    """ calculate Cells for splitting based on anchor points

    :param anchor_point_list:
    :param cell_size_meters:
    :return:
    cells_list [(minx, miny, maxx, maxy),
                ...]
    """
    cells_list = []
    for anchor_point in anchor_point_list:
        cells_list.append([anchor_point[0], anchor_point[1], anchor_point[0]+cell_size_meters, anchor_point[1]+cell_size_meters])

    return cells_list

def calculate_analysis_grid(utm_bounds, stride_size_meters=300, cell_size_meters=400):
    anchor_point_list = calculate_anchor_points(utm_bounds, stride_size_meters=stride_size_meters)

    cells_list = calculate_cells(anchor_point_list, cell_size_meters)

    return cells_list





if __name__ == '__main__':

    utmX, utmY = 658029, 4006947
    ll_x = utmX
    ur_x = utmX + 500
    ll_y = utmY
    ur_y = utmY + 500
    stride_size_meters = 300
    cell_size_meters = 400
    tile_size_pixels = 1600
    spacenetPath = "s3://spacenet-dataset/AOI_2_Vegas/srcData/rasterData/AOI_2_Vegas_MUL-PanSharpen_Cloud.tif"
    address = spacenetPath

    with rasterio.open(address) as src:

        wgs_bounds = utils.get_wgs84_bounds(src)
        utm_EPSG = utils.calculate_UTM_EPSG(wgs_bounds)
        utm_bounds = utils.get_utm_bounds(src, utm_EPSG)

        cells_list = calculate_analysis_grid(utm_bounds, stride_size_meters=stride_size_meters, cell_size_meters=cell_size_meters)

        random_cell = random.choice(cells_list)
        ll_x, ll_y, ur_x, ur_y = random_cell
        tile = tile_utm(src, ll_x, ll_y, ur_x, ur_y, indexes=None, tilesize=tile_size_pixels, nodata=None, alpha=None,
                 dst_crs=utm_EPSG)





